File: com.dynamo.cr.sceneed/src/com/dynamo/cr/sceneed/core/Camera.java
package com.dynamo.cr.sceneed.core;

import javax.vecmath.AxisAngle4d;
import javax.vecmath.Matrix4d;
import javax.vecmath.Point3d;
import javax.vecmath.Quat4d;
import javax.vecmath.Vector3d;
import javax.vecmath.Vector4d;

public class Camera {

    public enum Type {
        PERSPECTIVE, ORTHOGRAPHIC
    }

    private Vector3d position = new Vector3d();
    private Quat4d rotation = new Quat4d();

    private Matrix4d viewMatrix = new Matrix4d();
    private Matrix4d projection = new Matrix4d();
    private int[] viewport;

    private double zNear;
    private double zFar;
    private double aspect;

    private Type type;
    private double fov;

    public Camera(Type type) {
        this.type = type;
        this.zNear = 1;
        this.zFar = 2000;
        this.fov = 30;
        this.aspect = 1;
        double distance = 44.0;
        this.position.set(0.0, 0.0, 1.0);
        this.position.scale(distance);
        this.rotation.set(0.0, 0.0, 0.0, 1.0);
        reset();
    }

    public void reset() {
        this.viewport = new int[4];
        updateViewMatrix();
        updateProjectionMatrix();
    }

    public void setAspect(double aspect) {
        this.aspect = aspect;
        updateProjectionMatrix();
    }

    public double getAspect() {
        return aspect;
    }

    public void setFov(double fov) {
        this.fov = fov;
        updateProjectionMatrix();
    }

    public double getFov() {
        return fov;
    }

    public double getNearZ() {
        return zNear;
    }

    public double getFarZ() {
        return zFar;
    }

    public void setPerspective(double fov, double aspect, double znear,
            double zfar) {
        this.fov = fov;
        this.zNear = znear;
        this.zFar = zfar;
        this.aspect = aspect;
        this.type = Type.PERSPECTIVE;

        double xmin, xmax, ymin, ymax;

        ymax = znear * Math.tan(fov * Math.PI / 360.0);
        ymin = -ymax;
        xmin = ymin * aspect;
        xmax = ymax * aspect;

        setFrustum(xmin, xmax, ymin, ymax, znear, zfar);
    }

    public void setOrthographic(double fov, double aspect, double znear,
            double zfar) {
        this.fov = fov;
        this.zNear = znear;
        this.zFar = zfar;
        this.aspect = aspect;
        this.type = Type.ORTHOGRAPHIC;

        double left = -fov / 2;
        double right = fov / 2;

        double bottom = left / aspect;
        double top = right / aspect;

        setOrtho(left, right, bottom, top, znear, zfar);
    }

    public void setOrtho2D(double left, double right, double bottom, double top) {
        setOrtho(left, right, bottom, top, -1.0, 1.0);
    }

    public void setOrtho(double left, double right, double bottom, double top,
            double near, double far) {
        Matrix4d m = new Matrix4d();

        m.m00 = 2.0 / (right - left);
        m.m01 = 0.0;
        m.m02 = 0.0;
        m.m03 = -(right + left) / (right - left);

        m.m10 = 0.0;
        m.m11 = 2.0 / (top - bottom);
        m.m12 = 0.0;
        m.m13 = -(top + bottom) / (top - bottom);

        m.m20 = 0.0;
        m.m21 = 0.0;
        m.m22 = -2.0 / (far - near);
        m.m23 = -(far + near) / (far - near);

        m.m30 = 0.0;
        m.m31 = 0.0;
        m.m32 = 0.0;
        m.m33 = 1.0;

        projection.set(m);
    }

    private void setFrustum(double left, double right, double bottom,
            double top, double nearval, double farval) {
        double x, y, a, b, c, d;
        Matrix4d m = new Matrix4d();

        x = (2.0 * nearval) / (right - left);
        y = (2.0 * nearval) / (top - bottom);
        a = (right + left) / (right - left);
        b = (top + bottom) / (top - bottom);
        c = -(farval + nearval) / (farval - nearval);
        d = -(2.0 * farval * nearval) / (farval - nearval);

        m.m00 = x;
        m.m01 = 0.0;
        m.m02 = a;
        m.m03 = 0.0;
        m.m10 = 0.0;
        m.m11 = y;
        m.m12 = b;
        m.m13 = 0.0;
        m.m20 = 0.0;
        m.m21 = 0.0;
        m.m22 = c;
        m.m23 = d;
        m.m30 = 0.0;
        m.m31 = 0.0;
        m.m32 = -1.0;
        m.m33 = 0.0;

        projection.set(m);
    }

    public void setViewport(int x, int y, int w, int h) {
        this.viewport[0] = x;
        this.viewport[1] = y;
        this.viewport[2] = w;
        this.viewport[3] = h;
    }

    public void getViewport(int[] view_port) {
        for (int i = 0; i < 4; ++i)
            view_port[i] = viewport[i];
    }

    private void updateViewMatrix() {
        Matrix4d m = new Matrix4d();
        m.setIdentity();
        m.set(rotation);

        Vector3d pos = new Vector3d(position);
        m.transpose();
        m.transform(pos);
        pos.negate();
        m.setColumn(3, pos.x, pos.y, pos.z, 1.0);

        viewMatrix.set(m);
    }

    private void updateProjectionMatrix() {
        switch (type) {
        case PERSPECTIVE:
            setPerspective(fov, aspect, zNear, zFar);
            break;

        case ORTHOGRAPHIC:
            setOrthographic(fov, aspect, zNear, zFar);
            break;
        default:
            assert (false);
        }
    }

    public void getViewMatrix(Matrix4d m) {
        m.set(viewMatrix);
    }

    public void getProjectionMatrix(Matrix4d m) {
        m.set(projection);
    }

    public double[] getProjectionMatrixArray() {
        return toGLMatrixArray(projection);
    }

    public double[] getViewMatrixArray() {
        Matrix4d m = new Matrix4d();
        getViewMatrix(m);
        return toGLMatrixArray(m);
    }

    public Point3d project(double objx, double objy, double objz) {
        Vector4d in = new Vector4d();
        in.x = objx;
        in.y = objy;
        in.z = objz;
        in.w = 1.0;

        Vector4d out = new Vector4d();
        Matrix4d proj = new Matrix4d();
        Matrix4d model = new Matrix4d();
        getProjectionMatrix(proj);
        getViewMatrix(model);

        model.transform(in, out);
        proj.transform(out, in);

        if (in.w == 0.0)
            throw new java.lang.ArithmeticException();

        in.x /= in.w;
        in.y /= in.w;
        in.z /= in.w;

        double[] ret = new double[3];

        ret[0] = viewport[0] + (1 + in.x) * viewport[2] / 2;
        ret[1] = viewport[1] + (1 + in.y) * viewport[3] / 2;
        ret[2] = (1 + in.z) / 2;

        // Transform from "OpenGL" to "screen" y
        ret[1] = (viewport[3] - viewport[0]) - ret[1] - 1;

        return new Point3d(ret);
    }

    public Vector4d unProject(double winx, double winy, double winz) {
        Vector4d in = new Vector4d();

        // Transform from "screen" to "OpenGL"
        winy = (viewport[3] - viewport[0]) - winy - 1;

        in.x = (winx - viewport[0]) * 2 / viewport[2] - 1.0;
        in.y = (winy - viewport[1]) * 2 / viewport[3] - 1.0;
        in.z = 2 * winz - 1.0;
        in.w = 1.0;

        Matrix4d proj = new Matrix4d();
        Matrix4d model = new Matrix4d();
        getProjectionMatrix(proj);
        getViewMatrix(model);

        Matrix4d A = new Matrix4d();
        A.mul(proj, model);
        A.invert();

        Vector4d out = new Vector4d();
        A.transform(in, out);

        if (out.w == 0.0)
            throw new java.lang.ArithmeticException();

        return new Vector4d(out.x / out.w, out.y / out.w, out.z / out.w, 1);
    }

    public void rotate(Quat4d q) {
        Quat4d tmp = new Quat4d();
        tmp.set(q);
        tmp.normalize();
        rotation.mul(tmp);
        updateViewMatrix();
    }

    public void rotate(AxisAngle4d a) {
        Quat4d q = new Quat4d();
        q.set(a);
        rotate(q);
    }

    public void getRotation(Quat4d r) {
        r.set(rotation);
    }

    public void setRotation(Quat4d rotation) {
        rotation.set(rotation);
        updateViewMatrix();
    }

    public void move(double x, double y, double z) {
        position.x += x;
        position.y += y;
        position.z += z;
        updateViewMatrix();
    }

    private static double[] toGLMatrixArray(Matrix4d in) {
        Matrix4d m = new Matrix4d();
        m.set(in);
        m.transpose();

        double[] ret = new double[16];
        int i = 0;
        ret[i++] = m.m00;
        ret[i++] = m.m01;
        ret[i++] = m.m02;
        ret[i++] = m.m03;

        ret[i++] = m.m10;
        ret[i++] = m.m11;
        ret[i++] = m.m12;
        ret[i++] = m.m13;

        ret[i++] = m.m20;
        ret[i++] = m.m21;
        ret[i++] = m.m22;
        ret[i++] = m.m23;

        ret[i++] = m.m30;
        ret[i++] = m.m31;
        ret[i++] = m.m32;
        ret[i++] = m.m33;

        return ret;
    }

    public void setPosition(double x, double y, double z) {
        this.position.x = x;
        this.position.y = y;
        this.position.z = z;

        updateViewMatrix();
    }

    public Vector4d getPosition() {
        return new Vector4d(position.x, position.y, position.z, 1.0);
    }

    public Type getType() {
        return type;
    }
}

File: editor/src/clj/dynamo/camera.clj
(ns dynamo.camera
  (:require [dynamo.geom :as geom]
            [dynamo.graph :as g]
            [dynamo.types :as t]
            [dynamo.ui :as ui]
            [schema.core :as s])
  (:import [dynamo.types Camera Region AABB]
           [javax.vecmath Point3d Quat4d Matrix4d Vector3d Vector4d AxisAngle4d]))

(s/defn camera-view-matrix :- Matrix4d
  [camera :- Camera]
  (let [pos (Vector3d. (t/position camera))
        m   (Matrix4d.)]
    (.setIdentity m)
    (.set m (t/rotation camera))
    (.transpose m)
    (.transform m pos)
    (.negate pos)
    (.setColumn m 3 (.x pos) (.y pos) (.z pos) 1.0)
    m))

(s/defn camera-perspective-projection-matrix :- Matrix4d
  [camera :- Camera]
  (let [near   (.z-near camera)
        far    (.z-far camera)
        aspect (.aspect camera)
        fov    (.fov camera)

        ymax (* near (Math/tan (/ (* fov Math/PI) 360.0)))
        ymin (- ymax)
        xmin (* ymin aspect)
        xmax (* ymax aspect)

        x    (/ (* 2.0 near) (- xmax xmin))
        y    (/ (* 2.0 near) (- ymax ymin))
        a    (/ (+ xmin xmax) (- xmax xmin))
        b    (/ (+ ymin ymax) (- ymax ymin))
        c    (/ (- (+     near far)) (- far near))
        d    (/ (- (* 2.0 near far)) (- far near))
        m    (Matrix4d.)]
    (set! (. m m00) x)
    (set! (. m m01) 0.0)
    (set! (. m m02) a)
    (set! (. m m03) 0.0)

    (set! (. m m10) 0.0)
    (set! (. m m11) y)
    (set! (. m m12) b)
    (set! (. m m13) 0.0)

    (set! (. m m20) 0.0)
    (set! (. m m21) 0.0)
    (set! (. m m22) c)
    (set! (. m m23) d)

    (set! (. m m30) 0.0)
    (set! (. m m31) 0.0)
    (set! (. m m32) -1.0)
    (set! (. m m33) 0.0)
    m))

(s/defn camera-orthographic-projection-matrix :- Matrix4d
  [camera :- Camera]
  (let [near   (.z-near camera)
        far    (.z-far camera)
        right  (/ (.fov camera) 2.0)
        left   (- right)
        bottom (/ left (.aspect camera))
        top    (/ right (.aspect camera))

        m      (Matrix4d.)]
    (set! (. m m00) (/ 2.0 (- right left)))
    (set! (. m m01) 0.0)
    (set! (. m m02) 0.0)
    (set! (. m m03) (/ (- (+ right left)) (- right left)))

    (set! (. m m10) 0.0)
    (set! (. m m11) (/ 2.0 (- top bottom)))
    (set! (. m m12) 0.0)
    (set! (. m m13) (/ (- (+ top bottom)) (- top bottom)))

    (set! (. m m20) 0.0)
    (set! (. m m21) 0.0)
    (set! (. m m22) (/ -2.0 (- far near)))
    (set! (. m m23) (/ (- (+ far near) (- far near))))

    (set! (. m m30) 0.0)
    (set! (. m m31) 0.0)
    (set! (. m m32) 0.0)
    (set! (. m m33) 1.0)
    m))

(s/defn camera-projection-matrix :- Matrix4d
  [camera :- Camera]
  (case (:type camera)
    :perspective  (camera-perspective-projection-matrix camera)
    :orthographic (camera-orthographic-projection-matrix camera)))



(s/defn make-camera :- Camera
  ([] (make-camera :perspective))
  ([t :- (t/enum :perspective :orthographic)]
    (let [distance 10000.0
          position (doto (Point3d.) (.set 0.0 0.0 1.0) (.scale distance))
          rotation (doto (Quat4d.)   (.set 0.0 0.0 0.0 1.0))]
      (t/->Camera t position rotation 1 2000 1 30 (Vector4d. 0 0 0 1.0) (t/->Region 0 0 0 0)))))

(s/defn set-orthographic :- Camera
  [camera :- Camera fov :- t/Num aspect :- t/Num z-near :- t/Num z-far :- t/Num]
  (assoc camera
         :fov fov
         :aspect aspect
         :z-near z-near
         :z-far z-far))

(s/defn camera-rotate :- Camera
  [camera :- Camera q :- Quat4d]
  (assoc camera :rotation (doto (Quat4d. (t/rotation camera)) (.mul (doto (Quat4d. q) (.normalize))))))

(s/defn camera-move :- Camera
  [camera :- Camera x :- t/Num y :- t/Num z :- t/Num]
  (assoc camera :position (doto (Point3d. x y z) (.add (t/position camera)))))

(s/defn camera-set-position :- Camera
  [camera :- Camera x :- t/Num y :- t/Num z :- t/Num]
  (assoc camera :position (Point3d. x y z)))

(s/defn camera-set-center :- Camera
  [camera :- Camera bounds :- AABB]
  (let [center (geom/aabb-center bounds)
        view-matrix (camera-view-matrix camera)]
    (.transform view-matrix center)
    (set! (. center z) 0)
    (.transform ^Matrix4d (geom/invert view-matrix) center)
    (camera-set-position camera (.x center) (.y center) (.z center))))

(s/defn camera-project :- Point3d
  "Returns a point in device space (i.e., corresponding to pixels on screen)
   that the given point projects onto. The input point should be in world space."
  [camera :- Camera region :- Region point :- Point3d]
  (let [proj  (camera-projection-matrix camera)
        model (camera-view-matrix camera)
        in    (Vector4d. (.x point) (.y point) (.z point) 1.0)
        out   (Vector4d.)]
    (.transform model in out)
    (.transform proj out in)

    (assert (not= 0.0 (.w in)))
    (let [w (.w in)
          x (/ (.x in) w)
          y (/ (.y in) w)
          z (/ (.z in) w)

          rx (+ (.left region) (/ (* (+ 1 x) (.right  region)) 2))
          ry (+ (.top  region) (/ (* (+ 1 y) (.bottom region)) 2))
          rz (/ (+ 1 z) 2)

          device-y (- (.bottom region) (.top region) ry 1)]
      (Point3d. rx device-y rz))))

(defmacro scale-to-doubleunit [x x-min x-max]
  `(- (/ (* (- ~x ~x-min) 2) ~x-max) 1.0))

(defmacro normalize-vector [v]
  `(do
     (if (= (.w ~v) 0.0) (throw (ArithmeticException.)))
     (Vector4d. (/ (.x ~v) (.w ~v))
                (/ (.y ~v) (.w ~v))
                (/ (.z ~v) (.w ~v))
                1.0)))

(s/defn camera-unproject :- Vector4d
  [camera :- Camera win-x :- t/Num win-y :- t/Num win-z :- t/Num]
  (let [viewport (t/viewport camera)
        win-y    (- (.bottom viewport) (.top viewport) win-y 1.0)
        in       (Vector4d. (scale-to-doubleunit win-x (.left viewport) (.right viewport))
                            (scale-to-doubleunit win-y (.top viewport)  (.bottom viewport))
                            (- (* 2 win-z) 1.0)
                            1.0)
        proj     (camera-projection-matrix camera)
        model    (camera-view-matrix camera)
        a        (Matrix4d.)
        out      (Vector4d.)]
    (.mul a proj model)
    (.invert a)
    (.transform a in out)

    (normalize-vector out)))

(s/defn viewproj-frustum-planes :- [Vector4d]
  [camera :- Camera]
  (let [view-proj   (doto (camera-projection-matrix camera)
                      (.mul (camera-view-matrix camera)))
        persp-vec   (let [x (Vector4d.)] (.getRow view-proj 3 x) x)
        rows        (mapcat #(repeat 2 %) (range 3))
        scales      (take 6 (cycle [-1 1]))]
    (map (fn [row scale]
           (let [temp ^Vector4d (Vector4d.)]
             (.getRow view-proj ^int row temp)
             (.scale  temp scale)
             (.add    temp persp-vec)
             temp))
         rows scales)))

(defn dolly
  [camera delta]
  (update-in camera [:fov]
             (fn [fov]
               (max 0.01 (+ (or fov 0) (* (or fov 1) delta))))))

(defn track
  [camera last-x last-y evt-x evt-y]
  (let [focus ^Vector4d (:focus-point camera)
        point (camera-project camera (:viewport camera) (Point3d. (.x focus) (.y focus) (.z focus)))
        world (camera-unproject camera evt-x evt-y (.z point))
        delta (camera-unproject camera last-x last-y (.z point))]
    (.sub delta world)
    (assoc (camera-move camera (.x delta) (.y delta) (.z delta))
           :focus-point (doto focus (.add delta)))))

(defn tumble
  [^Camera camera last-x last-y evt-x evt-y]
  (let [rate 0.005
        dx (- last-x evt-x)
        dy (- last-y evt-y)
        focus ^Vector4d (:focus-point camera)
        delta ^Vector4d (doto (Vector4d. ^Point3d (:position camera))
                          (.sub focus))
        q-delta ^Quat4d (doto (Quat4d.)
                          (.set delta))
        r ^Quat4d (Quat4d. ^Quat4d (:rotation camera))
        inv-r ^Quat4d (doto (Quat4d. r)
                     (.conjugate))
        m ^Matrix4d (doto (Matrix4d.)
                    (.setIdentity)
                    (.set r)
                    (.transpose))
        q2 ^Quat4d (doto (Quat4d.)
                     (.set (AxisAngle4d. 1.0  0.0  0.0 (* dy rate))))
        y-axis ^Vector4d (doto (Vector4d.))
        q1 ^Quat4d (doto (Quat4d.))]
    (.mul q-delta inv-r q-delta)
    (.mul q-delta r)
    (.getColumn m 1 y-axis)
    (.set q1 (AxisAngle4d. (.x y-axis) (.y y-axis) (.z y-axis) (* dx rate)))
    (.mul q1 q2)
    (.normalize q1)
    (.mul r q1)
    ; rotation is now in r
    (.conjugate inv-r r)

    (.mul q-delta r q-delta)
    (.mul q-delta inv-r)
    (.set delta q-delta)
    (.add delta focus)
    ; position is now in delta
    (assoc camera
           :position (Point3d. (.x delta) (.y delta) (.z delta))
           :rotation r)))

(def ^:private button-interpretation
  {[:one-button 1 false true false]          :tumble
   [:one-button 1 true  true false]          :track
   [:one-button 1 true false false]          :dolly
   [:three-button 1 false true false]        :tumble
   [:three-button 2 false true false]        :track
   [:three-button 3 false true false]        :dolly})

(defn camera-movement
  ([event]
    (camera-movement (ui/mouse-type) (:button event) (:ctrl-down event) (:alt-down event) (:shift-down event)))
  ([mouse-type button control alt shift]
    (button-interpretation [mouse-type button control alt shift] :idle)))


(s/defn camera-fov-from-aabb :- t/Num
  [camera :- Camera ^AABB aabb :- AABB]
  (assert camera "no camera?")
  (assert aabb   "no aabb?")
  (let [^Region viewport (:viewport camera)
        min-proj    (camera-project camera viewport (.. aabb min))
        max-proj    (camera-project camera viewport (.. aabb max))
        proj-width  (Math/abs (- (.x max-proj) (.x min-proj)))
        proj-height (Math/abs (- (.y max-proj) (.y min-proj)))
        factor-x    (Math/abs (/ proj-width  (- (.right viewport) (.left viewport))))
        factor-y    (Math/abs (/ proj-height (- (.top viewport) (.bottom viewport))))
        factor-y    (* factor-y (:aspect camera))
        fov-x-prim  (* factor-x (:fov camera))
        fov-y-prim  (* factor-y (:fov camera))
        y-aspect    (/ fov-y-prim (:aspect camera))]
    (* 1.1 (Math/max y-aspect fov-x-prim))))

(s/defn camera-orthographic-frame-aabb :- Camera
  [camera :- Camera ^AABB aabb :- AABB]
  (assert (= :orthographic (:type camera)))
  (-> camera
    (set-orthographic (camera-fov-from-aabb camera aabb) (:aspect camera) (:z-near camera) (:z-far camera))
    (camera-set-center aabb)))

(g/defnode CameraController
  (property camera Camera)

  (property ui-state t/Any (default (constantly (atom {:movement :idle}))))
  (property movements-enabled t/Any (default #{:dolly :track :tumble}))

  (on :mouse-down
    (swap! (:ui-state self) assoc
      :last-x (:x event)
      :last-y (:y event)
      :movement (get (:movements-enabled self) (camera-movement event) :idle)))

  (on :mouse-move
    (let [{:keys [movement last-x last-y]} @(:ui-state self)
          {:keys [x y]} event]
      (when (not (= :idle movement))
        (case movement
          :dolly  (g/update-property self :camera dolly  (* -0.002 (- y last-y)))
          :track  (g/update-property self :camera track  last-x last-y x y)
          :tumble (g/update-property self :camera tumble last-x last-y x y)
          nil)
        (swap! (:ui-state self) assoc
          :last-x x
          :last-y y))))

  (on :mouse-up
    (swap! (:ui-state self) assoc
      :last-x nil
      :last-y nil
      :movement :idle))

  (on :mouse-wheel
    (when (contains? (:movements-enabled self) :dolly)
      (g/update-property self :camera dolly (* -0.02 (:count event))))))
